First PR for 2021 Design Team (#124)
ietf-wg-masque/draft-ietf-masque-h3-datagram: draft-ietf-masque-h3-datagram.md

congestion-control properties of QUIC. However, QUIC DATAGRAM frames do not
provide a means to demultiplex application contexts. This document describes how
to use QUIC DATAGRAM frames when the application protocol running over QUIC is
HTTP/3. It associates datagrams with client-initiated bidirectional streams.

> Additionally, this document defines how to convey datagrams over prior versions

Author: Agreed, done.

> # Multiplexing

Reviewer: Strictly editorial: a better way to structure this section would be to describe what we defin multiplexing is. Then the follow on paragraphs state how each version can (or cannot) satisfy that. That would mean simplifying paragraph 2 of this section by extracting the "need the ability" piece so that the text focuses on how not why.

Such an editorial change could be tracked as follow on work, since this PR is focuesd on design changes.

Author: This sounds good, but I agree - let's tackle this as editors once the outcome of the design team is merged.

> All HTTP Datagrams are associated with an HTTP request stream.

Reviewer: request stream, or simply request?

```suggestion
All HTTP Datagrams are associated with an HTTP request.
```

# Multiplexing

When running over HTTP/3, multiple exchanges of datagrams need the ability to
coexist on a given QUIC connection. To allow this, the QUIC DATAGRAM frame

> payload starts with an encoded stream identifier that associates the datagram
> with a given QUIC stream.

Reviewer 2: ```suggestion
payload starts with an encoded stream identifier that associates the datagram
with a request stream.
```

You already say "encoded stream identifier" and the goal is to associate with request streams.  In WebTransport, we might use those bidirectional streams without making them requests, which will require (or at least allow) some additional checking to see that the identified stream contains a request, but that's something WebTransport can handle.

When running over HTTP/2, demultiplexing is provided by the HTTP/2 framing

> layer. When running over HTTP/1, requests are strictly serialized in the

Author: Agreed, done.

> Receipt of a frame that includes a larger value MUST be treated as a connection

Reviewer 2: Noticed in reviewing the whole...  The line above says `(2<sup>62</sup>-1) / 4` where `2<sup>60</sup>-1` would do.

Author: Agreed, made the change.

This specification introduces the Capsule Protocol. The Capsule Protocol is a
sequence of type-length-value tuples that allows endpoints to reliably
communicate request-related information end-to-end on HTTP request streams, even
in the presence of HTTP intermediaries.

> ## Capsule Protocol {#capsule-protocol}

Reviewer 2: There is text missing from the introduction here that establishes the conditions on the use of capsule protocol.  That is, it only applies to those protocols that use extended CONNECT and also opt in to the use of the capsule protocol.

The nice thing with the draft up until this point is that it doesn't constrain the use of datagrams to those protocols that use extended CONNECT.  This part does.

(I know that this is perhaps fanciful, but I can imagine uses for regular HTTP requests that are enhanced by the use of DATAGRAM frames.)

Author: Agreed. I refactored the capsule section to contain more information up front.

> # Capsules {#capsule}

Reviewer: Editorial: Not strictly related to this PR but consider renaming this section to `Capsule Protocol` and the subsection to something else (possibly `Using the Capsule Protocol")

Author: Let's tackle that post design team

This specification introduces the Capsule Protocol. The Capsule Protocol is a
sequence of type-length-value tuples that new HTTP Methods or new HTTP Upgrade
Tokens can choose to use. It allows endpoints to reliably communicate
request-related information end-to-end on HTTP request streams, even in the
presence of HTTP intermediaries. The Capsule Protocol can be used to exchange

> HTTP Datagrams over versions of HTTP that do not support the QUIC DATAGRAM frame.

Reviewer: ambiguity alert: do we striclty mean HTTP versions? Presumably HTTP/3 connections that do not support QUIC DATAGRAM (or chose not to use it even if supported) are also in scope.

Consider moving the last sentence here to para 3, so that mention of Capsule Protocol plus HTTP Datagrams is grouped.

Author: Fixed

This specification defines the "data stream" of an HTTP request as the
bidirectional stream of bytes that follow the headers in both directions. In
HTTP/1.x, the data stream consists of all bytes on the connection that follow
the blank line that concludes either the request header section, or the 2xx
(Successful) response header section. In HTTP/2 and HTTP/3, the data stream of a
given HTTP request consists of all bytes sent in DATA frames with the

> corresponding stream ID. The concept of a data stream is particularly relevant

Reviewer: Editorial: you might be able to strengthen the definition of data stream a bit by extracting the semantic from version concerns. For example, something like

"This specification defines the "data stream" of an HTTP request as the
bidirectional stream of bytes that follow the header section in both directions. 
The concept of a data stream is particularly relevant for methods such as 
CONNECT. In the client-to-server direction, the data stream always exists.
In the server-to-client direction, the data stream only exists if when there
is a 2xx (Successful) response. In HTTP/1.x, the data stream consists of all
bytes on the connection that follow the blank line concluding header section.
In HTTP/2 and HTTP/3, the data stream of a given HTTP request consists
of all bytes sent in DATA frames with the corresponding stream ID. "

Author: Let's tackle that in a follow-up since the design team isn't touching the concept of the data stream

: The payload of this capsule. Its semantics are determined by the value of the
Capsule Type field.

Note that use of the Capsule Protocol is not required to use HTTP Datagrams.
Definitions of new HTTP Methods or of new HTTP Upgrade Tokens can use HTTP
Datagrams with their own data stream protocol.

> 

Reviewer 2: This bit belongs up front.

Author: Agreed. I refactored the capsule section to contain more information up front.

## Requirements

> 

Reviewer 2: > A server MUST NOT send any Transfer-Encoding or Content-Length header fields in a 2xx (Successful) response. If a client receives a Content-Length or Transfer-Encoding header fields in a successful response, it MUST treat that response as malformed.

If this is just about the response you need to say that:

1. the capsule protocol is not in use for a response unless it includes a 2xx status code
2. the capsule protocol MUST NOT be used with messages that contain Content-Length or Transfer-Encoding

You also say:

> A request message does not have content.

There are cases where Content-Length can be zero, but that is limited, not universal.  It's only CONNECT that gets that special handling.  For a new method, the capsule protocol will need to be used for the request body, which means that the request will have content.

> A successful response message does not have content.

Again, this is not going to be the case for anything other than CONNECT.

> Responses are not cacheable.

Generally true, but not necessarily universally true.  I think that this is more of a recommendation rather than a requirement.  It doesn't make sense to cache a response that uses the capsule protocol, but who are we to say that?

Author: Agreed, made the changes

: The payload of this capsule. Its semantics are determined by the value of the
Capsule Type field.

Capsules MUST be forwarded unmodified by intermediaries, with the exception of
the DATAGRAM capsule. An intermediary that understands the request semantics
enough to know that capsules are in use MAY parse, add, or remove DATAGRAM

> capsules; see {{datagram-capsule}}. Definitions of new Capsule Types MAY specify

Author: Agreed, tightened the text.

> optional custom intermediary processing.

Reviewer 3: This paragraph still concerns me. I'm happy to acknowledge that intermediaries exist, but I dont think we should be specifying their behavior. What happens if we just delete it?

Author: If we don't specify intermediary handling, intermediary implementors will implement what they feel like, such as dropping unknown capsules. If we do that we can't use capsules for end to end messages like we do in WebTransport and CONNECT-IP.

Reviewer 3: > If we don't specify intermediary handling, intermediary implementors will implement what they feel like, such as dropping unknown capsules. 

And the solution is for service operators not to use such intermediaries.

> If we do that we can't use capsules for end to end messages like we do in WebTransport and CONNECT-IP.

No, you can't do them in situations where you don't control the intermediaries.

This whole discussion of the position vis-a-vis intermediaries was a key point of disagreement on the call, so as long as that's unresolved, this PR isn't done.

Author: I'm not suggesting that this PR is done, resolving questions like this is why we're having this conversation :)

I see two options here:
1. the current text
2. remove mentions of intermediaries and require service operators terminate every protocol at the intermediary

I think (2) is more complex to deploy, and the cost of (1) is a few lines of spec that implementors of non-intermediaries can ignore, so I'm personally leaning towards (1).

Author: In today's design team meeting, we reached rough consensus to keep the current approach to intermediaries in this PR. We still have [an issue](https://github.com/ietf-wg-masque/draft-ietf-masque-connect-udp/issues/69) to track some changes that we still need to make, so I'm marking this comment thread as resolved.

use on responses unless the response includes a 2xx (Successful) status code.
The Capsule Protocol MUST NOT be used with messages that contain Content-Length
or Transfer-Encoding header fields.

## The DATAGRAM Capsule {#datagram-capsule}

This document defines the DATAGRAM capsule type (see {{iana-types}} for the
value of the capsule type). This capsule allows an endpoint to send an HTTP
Datagram on a stream using the Capsule Protocol. This is particularly useful
when using a version of HTTP that does not support QUIC DATAGRAM frames.

~~~
Datagram Capsule {
  Type (i) = DATAGRAM,
  Length (i),
  HTTP Datagram Payload (..),
}
~~~
{: #datagram-capsule-format title="DATAGRAM Capsule Format"}

HTTP Datagram Payload:

: The payload of the datagram, whose semantics are defined by individual
applications. Note that this field can be empty.

Datagrams sent using the DATAGRAM capsule have the same semantics as
datagrams sent in QUIC DATAGRAM frames. In particular, the restrictions on when
it is allowed to send an HTTP Datagram and how to process them from {{format}}
also apply to HTTP Datagrams sent and received using the DATAGRAM capsule.

> Intermediary can reencode HTTP Datagrams as it forwards them: in other words, an

Reviewer 4: ```suggestion
An intermediary can reencode HTTP Datagrams as it forwards them. In other words, an
```

: The payload of this capsule. Its semantics are determined by the value of the
Capsule Type field.

Capsules MUST be forwarded unmodified by intermediaries, with the exception of

> the DATAGRAM capsule. An intermediary that understands the request semantics

Reviewer: editorial: move `;see {{datagram-capsule}}` to the first ref here

Author: Done

enough to know that capsules are in use MAY convert between QUIC DATAGRAM frames
and DATAGRAM capsules; see {{datagram-capsule}}. Definitions of new Capsule
Types MAY specify optional custom intermediary processing.

Endpoints which receive a Capsule with an unknown Capsule Type MUST silently
drop that Capsule.

By virtue of the definition of the data stream, the Capsule Protocol is not in
use on responses unless the response includes a 2xx (Successful) status code.
The Capsule Protocol MUST NOT be used with messages that contain Content-Length

> or Transfer-Encoding header fields.

Reviewer: Editorial: Consider moving this paragraph closer to the definition of data stream, so that you can lose the first sentence.

Author: The purpose of this is to put requirements on the capsule protocol, so it wouldn't look right if moved to before the capsule protocol isn't introduced.

## The DATAGRAM Capsule {#datagram-capsule}

This document defines the DATAGRAM capsule type (see {{iana-types}} for the
value of the capsule type). This capsule allows an endpoint to send an HTTP
Datagram on a stream using the Capsule Protocol. This is particularly useful

> when using a version of HTTP that does not support QUIC DATAGRAM frames.

Reviewer: see earlier comment about whether this is strictly "version of HTTP" or not.

Author: Fixed

: In this mode, the intermediary terminates the data stream and parses all

> Capsule Type and Capsule Length fields it receives.

Reviewer 2: I think that the text on the two intermediary processing modes can be removed.

> Capsules MUST be forwarded unmodified by intermediaries, with the exception of
> the DATAGRAM capsule; the DATAGRAM capsule MAY be parsed, added, or removed by
> intermediaries, see {{datagram-capsule}}. Definitions of new Capsule Types MAY
> specify custom intermediary processing.

Reviewer 2: 👍

I might amend this to include the condition that the intermediary understand the semantics of the request enough to know that capsules are in use:

```suggestion
Capsules MUST be forwarded unmodified by intermediaries, with the exception of
the DATAGRAM capsule. An intermediary that understands the request semantics 
sufficient to know that capsules are in use MAY parse, add, or remove DATAGRAM
capsules; see {{datagram-capsule}}. Definitions of new Capsule Types MAY specify
custom intermediary processing.
```

What I don't want to see is intermediaries guessing that a particular unknown `:protocol` is using capsules when they do not.  Splicing in DATAGRAM frames/capsules in that context might go badly wrong.

> specify custom intermediary processing.

Reviewer 5: ```suggestion
specify optional custom intermediary processing.
```

If you have capsule types that cannot safely be passed verbatim by an intermediary, I don't think you're using The Capsule Protocol.

Author: Agreed, added optional.

Endpoints which receive a Capsule with an unknown Capsule Type MUST silently

> drop that Capsule.

Reviewer 2: This doesn't belong in intermediary processing.  Especially when it might be interpreted as applying to intermediaries.

Author: Agreed, moved the requirement on endpoints to the main capsule section

## The DATAGRAM Capsule {#datagram-capsule}

This document defines the DATAGRAM capsule type (see {{iana-types}} for the
value of the capsule type). This capsule allows an endpoint to send an HTTP
Datagram over an HTTP stream. This is particularly useful when using a version
of HTTP that does not support QUIC DATAGRAM frames.

~~~
Datagram Capsule {
  Type (i) = DATAGRAM,
  Length (i),
  HTTP Datagram Payload (..),
}
~~~
{: #datagram-capsule-format title="DATAGRAM Capsule Format"}

HTTP Datagram Payload:

: The payload of the datagram, whose semantics are defined by individual
applications. Note that this field can be empty.

Datagrams sent using the DATAGRAM capsule have the exact same semantics as
datagrams sent in QUIC DATAGRAM frames. In particular, the restrictions on when
it is allowed to send an HTTP Datagram and how to process them from {{format}}
also apply to HTTP Datagrams sent and received using the DATAGRAM capsule.

Intermediaries MAY parse DATAGRAM capsules and MAY send DATAGRAM capsules that

> they did not receive. This allows an intermediary to reencode HTTP Datagrams as

Reviewer 2: "MAY send DATAGRAM capsules that they did not receive" is problematic.  I don't think that we want intermediaries injecting their own special wisdom into the flow, even if that might be possible, it's not what we want to specifically allow.

What you are trying to say here is in the "in other words" clause: that an intermediary may translate QUIC DATAGRAM frames into HTTP DATAGRAM capsules and vice versa.

Author: Agreed, rephrased the text

it forwards them: in other words, an intermediary MAY send a DATAGRAM capsule to
forward an HTTP Datagram which was received in a QUIC DATAGRAM frame, and vice
versa.

> Note that while DATAGRAM capsules are sent on a stream, intermediaries can
> reencode DATAGRAM capsules into QUIC DATAGRAM frames over the next hop, and
> those could be dropped.

Reviewer 2: ```suggestion
Note that while DATAGRAM capsules that are sent on a stream are reliably delivered in order, intermediaries can
reencode DATAGRAM capsules into QUIC DATAGRAM frames when forwarding messages, which could result in loss or reordering.
```

the HTTP Datagram instead of converting it to a DATAGRAM capsule. This
preserves the end-to-end unreliability characteristic that methods such as
Datagram Packetization Layer Path MTU Discovery (DPLPMTUD) depend on
{{?RFC8899}}. An intermediary that converts QUIC DATAGRAM frames to DATAGRAM
capsules allows HTTP Datagrams to be arbitrarily large without suffering any

> loss; this can misrepresent the true path properties, defeating methods such a

Reviewer 6: ```suggestion
loss; this can misrepresent the true path properties, defeating methods such as
```

the HTTP Datagram instead of converting it to a DATAGRAM capsule. This
preserves the end-to-end unreliability characteristic that methods such as
Datagram Packetization Layer Path MTU Discovery (DPLPMTUD) depend on
{{?RFC8899}}. An intermediary that converts QUIC DATAGRAM frames to DATAGRAM
capsules allows HTTP Datagrams to be arbitrarily large without suffering any
loss; this can misrepresent the true path properties, defeating methods such a
DPLPMTUD.

## The RELIABLE_DATAGRAM Capsule

This document defines the RELIABLE_DATAGRAM capsule type (see {{iana-types}} for
the value of the capsule type). This capsule has the same format and semantics
as the DATAGRAM capsules, however intermediaries MUST NOT convert between
RELIABLE_DATAGRAM caspules and other encodings of HTTP Datagrams, or vice versa.
Endpoints MUST treat receipt of a RELIABLE_DATAGRAM capsule as equivalent to

> receipt of a DATAGRAM capsule with the same contents.

Reviewer 2: I'm fairly sure that we didn't agree to this bit.

On the contrary, I think that several of us opposed it.  Can we cut it and take it up separately?

Reviewer 7: I'm with Martin that I do not want to see this RELIABLE_DATAGRAM defined, especially not in this PR.

Especially given that capsules can't be modified by intermediaries, etc, I don't see the value in defining this here. If protocols want to define another capsule they use to transmit data over their stream of capsules, great—go ahead and define it there!

Author: I'd like us to discuss this in the context of this design team. The reason for that is that it has significant implications on future extensibility. I feel strongly that this spec needs enough of a toehold so that I can write a separate "datagram contexts" extension that has the "zero-latency extensibility" feature. The minimal toehold that I wrote initially was adding an UNEXTENDED_DATAGRAM capsule that has the same semantics as the DATAGRAM capsule but that intermediaries can't touch. Looking back I realized that this ended up being very similar to the RELIABLE_DATAGRAM that Ben asked for so I renamed it. I wrote up the way the extensibility toehold works in the "Creating a Demultiplexing Extension" appendix. I'm definitely open to other toehold that allow future extensibility, let me know if you have ideas.

Reviewer 8: I see some generic utility of RELIABLE_DATAGRAM.  It's, well, reliable, and is not bound by MTU restrictions.  The only difference between RELIABLE_DATAGRAM and an unknown Capsule type is that an endpoint has to treat it the same as receiving a DATAGRAM Capsule, instead of dropping it.

I'm missing how having it enables zero-latency extensibility.

Reviewer 3: I am also skeptical of this. 

I would ask if this question really needs to be resolved by this design team. Can't we discuss it in the WG?

Author: For me the extensibility properties are a fundamental part of the design, so the output of the design team needs to discuss this.

Reviewer 8: > this is unordered but reliable

How is it unordered if it can't be converted to a QUIC DATAGRAM?

> That said, I agree with Martin that it doesn't belong in this PR. Let's work on one piece at a time

+1 

There's a lot here we have agreement on, so if RELIABLE_DATAGRAM is controversial let's break it out.  

I'm still missing how it enables or is required for 0-RTT extensions (which are a nice-to-have).  The appendix was vague.  @DavidSchinazi can you write down what you are thinking.

Author: The current draft (version -05) contains the datagram context extensibility system. I'm OK with us removing that from this draft (which is what this PR does), on the condition that I can still implement that as an extension to this draft. In particular, I need that extension to work without costing a round trip (what I've been calling "zero-latency extensibility"). Let me go into more detail as to why RELIABLE_DATAGRAM is important to that.

The datagram context extension works as follows: it is negotiated by adding a header to both request and response, and when both endpoints support it, it adds a context ID varint at the start of the HTTP Datagram Payload field.

Now, in order to achieve zero-latency extensibility, the client needs to be able to send datagrams before receiving the server headers in such a way that the server can correctly parse them (if you add the context ID and the server doesn't support the extension, you don't want the server to parse the context ID as part of the payload).

So here's my idea for how to accomplish that. The extension says the following:
- clients that support the extension are not allowed to send QUIC DATAGRAM frames nor DATAGRAM capsule before they've received the server headers
- the extension defines a new RELIABLE_DATAGRAM_EXT capsule that can be sent before and after receiving server headers
```
Extended Reliable Datagram Capsule {
  Type (i) = RELIABLE_DATAGRAM_EXT,
  Length (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
```
- the extension does not change the semantics of the RELIABLE_DATAGRAM capsule, that still does not contain a Context ID
```
Reliable Datagram Capsule {
  Type (i) = RELIABLE_DATAGRAM,
  Length (i),
  HTTP Datagram Payload (..),
}
```
- when both endpoints support the extension (after receiving headers) the extension changes the semantics of both the QUIC DATAGRAM frame and DATAGRAM capsule to include a Context ID at the start:
```
Datagram Capsule {
  Type (i) = DATAGRAM,
  Length (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
HTTP/3 Datagram { // contents of QUIC DATAGRAM frame
  Quarter Stream ID (i),
  Context ID (i),
  HTTP Datagram Payload (..),
}
```
- if the client wishes to use zero-latency extensibility, it can send RELIABLE_DATAGRAM and RELIABLE_DATAGRAM_EXT before receiving the server headers. Both of these will be parsed correctly by servers, whether or not the server supports the extension.

All of this relies on the fact that there is one bit that differentiates the DATAGRAM capsule from the RELIABLE_DATAGRAM capsule (the bit is in the capsule type), so that the extension can give the two capsules different semantics in a way that is backwards compatible.

What I like about this design is that the added complexity to the HTTP Datagram spec is as minimal as possible: it just requires two case statements instead of one in the switch statement that parses capsule types.

I'm OK removing the RELIABLE_DATAGRAM capsule from this PR, as long as we have some other mechanism that I can use to write my extension.

Reviewer 9: I find this entire conversation a bit hard to follow, so forgive me if I've missed anything.

There's a claim that RELIABLE_DATAGRAM is controversial, and also a poor name. I haven't seen a good argument to remove it, merely the assertion that people do not want it. There are however a few use cases that have been described in its favor, so I personally would like to see it included (even if it has some other name), as per the discussion we've had at IETF and on list around the different MTU issues, as well as the H3->H2->H3 translation issues that it helps resolve.

A different issue that I'd like to raise (although I can live with this) is that I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here. I understand that is a natural consequence of the Datagram Capsule -> QUIC DATAGRAM encoding, but I'd personally really like to see the type varint make it into the QUIC DATAGRAM so that we can always unambiguously parse the datagrams, regardless of how we received them (capsule or otherwise). I acknowledge that this may not be the way others want this, and can live with the current (proposed) text.

Reviewer 3: > There's a claim that RELIABLE_DATAGRAM is controversial, and also a poor name. I haven't seen a good argument to remove it, merely the assertion that people do not want it. 

This is the wrong standard. Things need consensus to be *included*, not removed.

In any case, a comment on this piece of the document is the wrong place to debate that, so the right thing to do is not have it in this PR and then discuss it separately.

Reviewer 9: I disagree with your position that it is reasonable to omit things just because someone provides an unsubstantiated objection. Kindly please provide a technical argument for its removal, or we will not be able to build consensus in either direction.

Reviewer 8: Maybe I don't understand what 0-RTT extensibility means.  Does it mean:

1. A client that supports an extension can try to use the extension before receiving confirmation that the server supports it
2. If the server supports the extension it will be processed correctly
3. If the server does not support the extension, the stream will not be torn down, but the extended datagrams could be dropped
---
If that's correct, can you build this without RELIABLE_DATAGRAM by:

1. Extension defines new Capsule type DATAGRAM_EXT
2. Client sends indication that it supports DATAGRAM_EXT in the CONNECT headers
3. Client sends DATAGRAM_EXT Capsules
3a.  Server supports extension, processes DATAGRAM_EXT Capsules
4a. Server confirms extension support
5a. Client receives acknowledgement that the extension is supported, switches to DATAGRAM and DATAGRAM Capsules
3b. Server silently drops DATAGRAM_EXT Capsules
4b. Server sends CONNECT response
5b. Client sees the server does not support the extension, uses standard DATAGRAM and DATAGRAM Capsules

I see that RELIABLE_DATAGRAM makes it possible for a server that supports the extension to assert than any DATAGRAM or DATAGRAM Capsule is only the extended flavor.  But how would an extended server handle a non-extended client in that case?  It seems like the extended server has to have a switch to know if the DATAGRAMs are extended or not, so I don't see any value in RELIABLE_DATAGRAM in this use case.  

Please help me understand if I'm missing something.

Reviewer 3: The way this process is supposed to work is that *this* PR is supposed to include the points on which we had consensus, and that means when people object to things which clearly don't, they are removed and deferred for later discussion. If that's not going to happen, then we can just abandon this PR and go back to real time discussion.

Reviewer 8: > A different issue that I'd like to raise (although I can live with this) is that I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here.

I am inclined to agree, but I see the solution as moving the demux higher in the protocol stack (eg: to CONNECT-UDP or CONNECT-IP) rather than down (into QUIC or H3 Datagram).  CONNECT-IP can say all DATAGRAMs on a CONNECT-IP stream have a specific format after the stream ID that includes "Context ID", "Format Type", etc, and this doesn't need to be negotiated at all.  I'd slightly prefer if those protocols used less generic concepts in doing so, but can live with it either way.

Reviewer 9: ekr: You appear to have a particularly rigid definition of "the way this is supposed to work", and it does not match mine, and I rather suspect many other people's. I am not particularly interested in having a procedural discussion here, but rather working towards consensus on the document. To that end, I would like to see technical reasoning for everything that goes _in_ or gets taken _out_. I do not consider emotional appeals such as "I don't want this" to be valid.

Reviewer 9: afrind: The one downside with your suggestion is it means the context sensitivity still exists, but is a property of the protocol rather than an extension. I think that makes logical sense, but doesn't fully solve the issue I was describing. It's probably the case that we don't necessarily have to solve it, as I mentioned before, I can live with the semantic change here.

Author: @afrind regarding your proposal in https://github.com/ietf-wg-masque/draft-ietf-masque-h3-datagram/pull/124#discussion_r753515209 the issue there is that it doesn't allow the client to send an unextended datagram in its first flight. For CONNECT-UDP that's really useful: the client wants to send its first UDP payload with the first flight and have the server still correctly parse that even if it doesn't support the extension.

@ekr and @achernya in terms of process here: it's clear that there isn't consensus to land this PR with RELIABLE_DATAGRAM, and there isn't consensus to land it without RELIABLE_DATAGRAM. We hadn't really discussed RELIABLE_DATAGRAM at our last in-person design team so I wasn't sure what people's thoughts on it were. Clearly it's something that warrants in-person discussion time. My plan is to keep the topic of RELIABLE_DATAGRAM as an open issue that we will resolve at our next in-person design team meeting. This PR won't get merged until this is resolved.

Reviewer 8: > the issue there is that it doesn't allow the client to send an unextended datagram in its first flight.

I assume you mean 'unextended DATAGRAM Capsule'.  In that case, I think I understand, which is that a DATAGRAM Capsule can be converted to a QUIC DATAGRAM by an intermediary, and there's ambiguity at a server that supports the extension after it confirms support: when did the client receive confirmation and start sending extended QUIC DATAGRAMs?

The client can still send DATAGRAM Capsules in this case, and move to QUIC DATAGRAM after it knows whether the server supports the extension or not.  Is there a real use case for sending both extended and non-extended DATAGRAMs in 0-RTT, where we're willing to sacrifice HoL blocking resilience and possible flow control constraints for the extended ones, but not willing to do so for the non-extended ones?  Or maybe the use-cases are small but we think the cost of RELIABLE_DATAGRAM is low and maybe has other utility?

This brings me back here:

> I'm not really a fan of the extension-that-changes-the-semantics-of-DATAGRAM scheme that is being proposed here.

I think we've shown that it's pretty hard to negotiate changes to QUIC DATAGRAM formats on the fly.  Let's not do that.  The cost of easy extensibility is an additional byte (or two) of header information inside the datagram payload of protocols that need it.  If CONNECT-UDP needs to have a such a header I can live with that.

Author: QUIC DATAGRAM frames and DATAGRAM capsules need to have the exact same semantics, because intermediaries convert blindly between them so you can't know how they were originally sent. That's why my proposal forbids sending either when you support the extension and haven't yet received the peer's headers.

My personal opinion is that
- the cost of having RELIABLE_DATAGRAM for an implementation that doesn't care is very low
- RELIABLE_DATAGRAM has value in and of itself
- moving the complexity of contexts to their own extension allows defining different ways to do this down the road, whereas if we say that CONNECT-UDP starts with a varint we've committed to this context design forever

Let's discuss this in person on 11/29

Reviewer 2: Congratulations, you have just collectively managed to do what we've learned through hard experience is a bad idea: a substantive technical discussion in a comment attached to a line of code on a pull request.

All I was asking for was for this to be taken to an issue.  Can we do that?

Author: Sure! I've filed #125 to discuss this.

Author: I've removed RELIABLE_DATAGRAM from this PR. We still have #111 to discuss RELIABLE_DATAGRAM and #125 to discuss zero-latency extensibility, so I'm now marking this comment thread as resolved.

the HTTP Datagram instead of converting it to a DATAGRAM capsule. This
preserves the end-to-end unreliability characteristic that methods such as
Datagram Packetization Layer Path MTU Discovery (DPLPMTUD) depend on
{{?RFC8899}}. An intermediary that converts QUIC DATAGRAM frames to DATAGRAM
capsules allows HTTP Datagrams to be arbitrarily large without suffering any
loss; this can misrepresent the true path properties, defeating methods such a
DPLPMTUD.

## The RELIABLE_DATAGRAM Capsule

This document defines the RELIABLE_DATAGRAM capsule type (see {{iana-types}} for
the value of the capsule type). This capsule has the same format and semantics
as the DATAGRAM capsules, however it is opaque to intermediaries: intermediaries
MUST NOT convert between QUIC DATAGRAM frames and RELIABLE_DATAGRAM capsules or

> vice versa. Endpoints MUST treat receipt of a RELIABLE_DATAGRAM capsule as

Reviewer: what about RELIABLE_DATAGRAM -> DATAGRAM capsule conversion?

Maybe simpler to just state that they are passed through and MUST NOT be converted to anything else.

endpoints negotiate the same draft version.

> # The Capsule-Protocol Header

Reviewer: Editorial: consider moving this whole section to be part of the definition of Capsule protocol definition section

Author: Done

Definitions of new HTTP Methods or of new HTTP Upgrade Tokens that use the
Capsule Protocol MAY use the Capsule-Protocol header field to simplify
intermediary processing.

"Capsule-Protocol" is an Item Structured Header {{!RFC8941}}. Its value MUST be
a Boolean, its ABNF is:

~~~
Capsule-Protocol = sf-boolean
~~~

Endpoints indicate that the Capsule Protocol is in use on the data stream by
sending the Capsule-Protocol header field with a value of ?1. A Capsule-Protocol
header field with a value of ?0 has the same semantics as when the header is not
present. Intermediaries MAY use this header field to allow processing of HTTP
Datagrams for unknown HTTP methods or unknown HTTP Upgrade Tokens.

The Capsule-Protocol header field MUST NOT be sent multiple times on a message.
The Capsule-Protocol header field MUST NOT be used on HTTP responses with a

> status code different from 2xx (Successful). This specification does not define

Author: Let's take that to issues (I filed #130 and #131), I think we can handle this outside the design team

endpoints negotiate the same draft version.

# The Capsule-Protocol Header

Definitions of new HTTP Methods or of new HTTP Upgrade Tokens that use the
Capsule Protocol MAY use the Capsule-Protocol header field to simplify
intermediary processing.

"Capsule-Protocol" is an Item Structured Header {{!RFC8941}}. Its value MUST be
a Boolean, its ABNF is:

~~~

> Capsule-Protocol = sf-boolean

Reviewer 10: I am inclined to make this `Sec-`, because I don't see a scenario in which JavaScript would need to send it, and in general I believe that new request headers should be defined as `Sec-` unless they're both safe and useful for JavaScript applications.

Author: It sounds like we have folks who prefer with `Sec-` and folks who prefer without. Let's discuss in person.

Author: In today's design team meeting, we reached rough consensus to not use the `Sec-` prefix. We still have [an issue](https://github.com/ietf-wg-masque/draft-ietf-masque-connect-udp/issues/68) to track adding text to Security Considerations, so I'm marking this comment thread as resolved.

> methods or unknown HTTP Upgrade Tokens.

Reviewer 3: This is more material specifying intermediaries.

Author: That is correct.

"Sec-Capsule-Protocol" is an Item Structured Header {{!RFC8941}}. Its value MUST
be a Boolean, its ABNF is:

~~~

> Sec-Capsule-Protocol = sf-boolean

Reviewer 3: What are the semantics of this boolean? I don't see them defined.

Author: Good point, I've added text to clarify the semantics.

endpoints negotiate the same draft version.

> # The Sec-Capsule-Protocol Header

Reviewer 5: Why does this have a Sec- prefix?  Why shouldn't Javascript be allowed to set this header on requests it generates?  It seems safe to me.

Author: That sounds dangerous, if we allow JavaScript to send this we're adding a burden on every new capsule type to consider the implications of that. I'd rather disallow it to limit risk and avoid potential issues. If someone has a use-case that requires this we could reconsider that choice.

Reviewer 2: I know that we just shove "Sec-" on everything nowadays, but I think we need to meet a higher bar than "sounds dangerous".  Do you have a concrete example?

Let's say that I did a POST using fetch() with this new field attached.  I do that over h2 and a CDN upgrades to h3 toward the origin, splitting out apparent "datagrams" into frames on that leg.  The server might still accept the request, but discard the datagrams.  That means that the CDN and origin server see a different payload.  Maybe that's something that can be exploited.

But then, what is special about that being dependent on JavaScript?  I don't think that it works as an attack. Fetch will preflight the request and will not add the header unless it is specifically allowed by the server in a preflight, see [CORS-unsafe request-header names](https://fetch.spec.whatwg.org/#cors-unsafe-request-header-names).

Author: That's fair. You understand CORS better than I do. After thinking about this some more I have no strong opinion here so I'm happy to go with the consensus of the design team and/or WG on whether we should start with "Sec-" or not.

Author: I've removed the Sec- prefix.

Definitions of new HTTP Methods or of new HTTP Upgrade Tokens that use the

> Capsule Protocol MAY use the Sec-Capsule-Protocol header to simplify

Reviewer 2: ```suggestion
Capsule Protocol MAY use the Sec-Capsule-Protocol header field to simplify
```

...and throughout.

intermediary processing.

Endpoints indicate that the Capsule Protocol is in use on the data stream by
sending the Sec-Capsule-Protocol header with a value of ?1. Intermediaries MAY
use this header to allow processing of HTTP Datagrams for unknown HTTP methods

> or unknown HTTP Upgrade Tokens.

Reviewer 2: What if this appears on a response that doesn't have a 2xx status?  What about 1xx?

Author: Clarified that Sec-Capsule-Protocol is only allowed on 2xx

"Sec-Capsule-Protocol" is an Item Structured Header {{!RFC8941}}. Its value MUST

> be a Boolean, its ABNF is:

Reviewer 2: This bit where you can't define the field as a Boolean directly is weird.  @mnot, why not?

Also, I think that you need to define what to do with multiple values and parameters.

Reviewer 7: I think this is just the standard suggested text for structured fields 🤷

Author: Yes, this is standard boilerplate.

I added text about multiple values and parameters.

| Capsule Type                 |   Value   | Specification |
|:-----------------------------|:----------|:--------------|
| DATAGRAM                     | 0xff37a0  | This Document |

> | RELIABLE_DATAGRAM            | 0xff37a5  | This Document |

Reviewer: appreciate this is just the first proposal but the format changes and codepoint usage here seems breaking, sp probably requires a bump to the H3_DATAGRAM setting codepoint.

Author: Interestingly, these changes don't break backwards compatibility because the format is the same in the absence of the Sec-Use-Datagram-Contexts header. So we don't need to change the setting or codepoints - which is nice because it doesn't break existing WebTransport implementations.

Reviewer: Right but previously the intermediary could manipulate DATAGRAM capsules with type 0xff37a5, and now the text says they can't do that because the semantics of the RELIABLE_DATAGRAM are different. Is the assumption that no one implemented an intermediary that procesed DATAGRAM before this change?

Reviewer: actually, that's a problem with capsule protocol versioning not the H3 SETTING but I think my point stands

Author: Ah right, yes the (unstated) assumption was that no one has yet implemented an intermediary yet so we don't need to maintain backwards compatibility yet.

  :scheme = https
  :path = /target.example.org/443/
  :authority = proxy.example.org:443

>   sec-capsule-protocol = ?1

Reviewer: I note the response doesn't have the equivalent header. I assume that's intended but I'm unclear why based on the header's definition.

Author: No that was an oversight, fixed.

  :method = webtransport
  :path = /hello
  :authority = webtransport.example.org:443
  origin = https://www.example.org:443

>   sec-capsule-protocol = ?1

Reviewer 2: It would be good to have an example that uses capsules but does not include this header field.

Reviewer 7: Agreed

Author: I removed the header from the WebTransport example.
